### ISSUE TITLE
Archiving GRAALVM_CMP_GU_GVM_POLYNATIVE_RGX_SQ_STAGE1_SVM_TFL failed

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#616]
When trying to build native images with SubstrateVM, I always get the following error:

```bash
Process Process-172:
Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/process.py", line 267, in _bootstrap
    self.run()
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "graal/mx/mx.py", line 12177, in executeTask
    task.execute()
  File "graal/mx/mx.py", line 810, in execute
    self.build()
  File "graal/graal/vm/mx.vm/mx_vm.py", line 413, in build
    super(GraalVmLayoutDistributionTask, self).build()
  File "graal/mx/mx.py", line 1722, in build
    self.subject.make_archive()
  File "graal/mx/mx.py", line 2377, in make_archive
    self._install_source(source, output, destination, arc)
  File "graal/mx/mx.py", line 2314, in _install_source
    _install_source_files(((source_file, _arcname_f(source_file)) for source_file in glob.iglob(file_path)), include=source_path, excludes=source.get('exclude'))
  File "graal/mx/mx.py", line 2171, in _install_source_files
    merge_recursive(_source_file, _dst, _arcname, excludes)
  File "graal/mx/mx.py", line 2141, in merge_recursive
    merge_recursive(join(src, name), join(dst, name), join(src_arcname, name), excludes)
  File "graal/mx/mx.py", line 2141, in merge_recursive
    merge_recursive(join(src, name), join(dst, name), join(src_arcname, name), excludes)
  File "graal/mx/mx.py", line 2141, in merge_recursive
    merge_recursive(join(src, name), join(dst, name), join(src_arcname, name), excludes)
  File "graal/mx/mx.py", line 2141, in merge_recursive
    merge_recursive(join(src, name), join(dst, name), join(src_arcname, name), excludes)
  File "graal/mx/mx.py", line 2137, in merge_recursive
    os.symlink(link_target, absolute_destination)
OSError: [Errno 17] File exists
Archiving GRAALVM_CMP_GU_GVM_POLYNATIVE_RGX_SQ_STAGE1_SVM_TFL failed
error while killing subprocess 33242 "graalvm-ee-1.0.0-rc5/Contents/Home/bin/java -d64 -cp graal/mx/mxbuild/java/com.oracle.mxtool.compilerserver/bin:graalvm-ee-1.0.0-rc5/Contents/Home/lib/tools.jar com.oracle.mxtool.compilerserver.JavacDaemon -j 8": [Errno 3] No such process
1 build tasks failed
```

I'm not sure if this is a `SubstrateVM` or `mx` issue.

**Comment by ansalond** [comment]
Hello @fniephaus, thank you for the report. Unfortunately I'm not able to reproduce the issue.
I see that you are building a GraalVM image that includes the `sq` component. Is the corresponding repository open source?

In the meanwhile I'll try to provide a more meaningful error message.

**Comment by fniephaus** [comment]
Hi @ansalond,
No, the project is not open source yet, but I've sent you an invitation. Just run `mx squeak-svm` and you should be able to reproduce the problem on macOS.
Let me know if you need any further help.

**Comment by ansalond** [comment]
Thanks for the invitation, but I don't think I'm allowed to access your repository.
Anyway, the issue should be fixed in [the current tip of mx](https://github.com/graalvm/mx/commit/0bad96a0d809b13f3efac2b5b55595c8db3da1ec).

Just a curiosity: from the stack trace it looks like the `sq` component overwrites an existing file with a symlink. Is this intentional and necessary? Do you get a `Duplicate archive entry` warning?

**Comment by fniephaus** [comment]
Thanks, I'm waiting for CI to confirm this is working. And yes, I do get lots of "Duplicate archive entry" warnings actually. Not sure what is going on there, maybe caused by some misconfiguration? According to GitHub, you didn't accept the invitation, sending a new one now...

**Comment by fniephaus** [comment]
Unfortunately, this hasn't fixed the problem entirely...here's the new error copied from CI:

```bash
Archiving GRAALVM_CMP_GU_GVM_POLYNATIVE_RGX_SQ_STAGE1_SVM_TFL... [dependency LOCATOR updated]
WARNING:   GRAALVM_CMP_GU_GVM_POLYNATIVE_RGX_SQ_STAGE1_SVM_TFL:
Adding file which is not in the repository: '/home/travis/graalvm-ce-1.0.0-rc5' in 'graalvm-cmp-gu-gvm-polynative-rgx-sq-stage1-svm-tfl-1.0.0-rc5'
WARNING:   GRAALVM_CMP_GU_GVM_POLYNATIVE_RGX_SQ_STAGE1_SVM_TFL:
Duplicate archive entry: 'graalvm-cmp-gu-gvm-polynative-rgx-sq-stage1-svm-tfl-1.0.0-rc5/3rd_party_licenses.txt' previously added by graalvm-cmp-gu-gvm-polynative-rgx-sq-stage1-svm-tfl-1.0.0-rc5/./<-extracted-dependency:vm:VM_GRAALVM_SUPPORT added again by graalvm-cmp-gu-gvm-polynative-rgx-sq-stage1-svm-tfl-1.0.0-rc5<-file:/home/travis/graalvm-ce-1.0.0-rc5
WARNING:   GRAALVM_CMP_GU_GVM_POLYNATIVE_RGX_SQ_STAGE1_SVM_TFL:
Duplicate archive entry: 'graalvm-cmp-gu-gvm-polynative-rgx-sq-stage1-svm-tfl-1.0.0-rc5/bin/gu' previously added by graalvm-cmp-gu-gvm-polynative-rgx-sq-stage1-svm-tfl-1.0.0-rc5/bin/<-link:../jre/bin/gu added again by graalvm-cmp-gu-gvm-polynative-rgx-sq-stage1-svm-tfl-1.0.0-rc5<-file:/home/travis/graalvm-ce-1.0.0-rc5
Process Process-168:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/home/travis/build/mx/mx.py", line 12189, in executeTask
    task.execute()
  File "/home/travis/build/mx/mx.py", line 810, in execute
    self.build()
  File "/home/travis/build/graal/vm/mx.vm/mx_vm.py", line 413, in build
    super(GraalVmLayoutDistributionTask, self).build()
  File "/home/travis/build/mx/mx.py", line 1722, in build
    self.subject.make_archive()
  File "/home/travis/build/mx/mx.py", line 2379, in make_archive
    self._install_source(source, output, destination, arc)
  File "/home/travis/build/mx/mx.py", line 2325, in _install_source
    _install_source_files(((source_file, _arcname_f(source_file)) for source_file in glob.iglob(file_path)), include=source_path, excludes=source.get('exclude'))
  File "/home/travis/build/mx/mx.py", line 2182, in _install_source_files
    merge_recursive(_source_file, _dst, _arcname, excludes)
  File "/home/travis/build/mx/mx.py", line 2152, in merge_recursive
    merge_recursive(join(src, name), join(dst, name), join(src_arcname, name), excludes)
  File "/home/travis/build/mx/mx.py", line 2152, in merge_recursive
    merge_recursive(join(src, name), join(dst, name), join(src_arcname, name), excludes)
  File "/home/travis/build/mx/mx.py", line 2148, in merge_recursive
    add_symlink(src, link_target, absolute_destination, dst)
  File "/home/travis/build/mx/mx.py", line 2133, in add_symlink
    os.symlink(src, abs_dest)
OSError: [Errno 17] File exists
Archiving GRAALVM_CMP_GU_GVM_POLYNATIVE_RGX_SQ_STAGE1_SVM_TFL failed
1 build tasks failed
```

**Comment by ansalond** [comment]
It looks like the `JAVA_HOME` environment variable points to `/home/travis/graalvm-ce-1.0.0-rc5`. Unfortunately, you cannot yet use a GraalVM to build a new GraalVM. You need to use a JVMCI-enabled JDK8 like, e.g., one of those available here: https://github.com/graalvm/openjdk8-jvmci-builder/releases.

Does this solve your issue?

**Comment by fniephaus** [comment]
mh, I can successfully build a GraalSqueak native image with Substrate from GraalVM RC5 locally. But let me try again with a JVMCI-enabled JDK8 tomorrow...

**Comment by fniephaus** [comment]
BTW, I was also able to build on GraalVM RC5 with [this patch](https://github.com/fniephaus/mx/commit/506acee9960e9e3b02419f279c594280f1ac9aa7) applied to mx...

**Comment by ansalond** [comment]
During the GraalVM build process we copy files from the "base" JDK (e.g., the one `JAVA_HOME` points to) to the new GraalVM. If you bootstrap a GraalVM with a GraalVM, some unwanted files from the "old" one might end up in the "new" one. We plan support this use case in the future, but we cannot do it right now.
I will fix mx to avoid throwing the error, but I'd still recommend you to build from a plain JVMCI-enabled JDK rather than a GraalVM.

**Comment by ansalond** [comment]
I merged [some additional changes to mx](https://github.com/graalvm/mx/commit/7bb34ea43d03d86eef263311a88b25f166a38b3c) to support overwriting of broken symlinks. However, I still recommend you to use a plain JVMCI-enabled JDK to build GraalVM images.

**Comment by fniephaus** [comment]
Thanks! Building with a JVMCI-enabled JDK works like a charm now 😃

**Comment by ansalond** [comment]
Glad to hear that! Thanks for all the feedback!
